src/plugins/layout/nodeinfo.ts

```diff
@@ -1,50 +1,22 @@
-import { ILayoutState, INodeInfo } from "@clarity-types/layout";
-import { createLayoutState, getNodeIndex, Tags } from "./stateprovider";
+import { INodeInfo } from "@clarity-types/layout";
+import { shouldIgnoreNode } from "./states/ignore";
+import { isCssText } from "./states/text";
 
-export const ForceMaskAttribute = "data-clarity-mask";
+export const UnmaskAttribute = "data-clarity-unmask";
 
 export function createNodeInfo(node: Node, parentInfo: INodeInfo): INodeInfo {
-  let index = getNodeIndex(node);
-  let ignore = shouldIgnoreNode(node, parentInfo ? parentInfo.ignore : false);
-  let forceMask = shouldForceMaskNode(node, parentInfo ? parentInfo.forceMask : false);
-  let state: ILayoutState = createLayoutState(node, ignore, forceMask);
-  return { index, ignore, forceMask, state };
+  let ignore = shouldIgnoreNode(node, parentInfo);
+  let unmask = shouldUnmaskNode(node, parentInfo);
+  let isCss = isCssText(node);
+  return { ignore, unmask, isCss };
 }
 
-export function shouldIgnoreNode(node: Node, parentIgnored: boolean): boolean {
-  let ignore = parentIgnored;
-  if (!parentIgnored) {
-    switch (node.nodeType) {
-      case Node.ELEMENT_NODE:
-        let tagName = (node as Element).tagName;
-        if (tagName === Tags.Script || tagName === Tags.Meta) {
-          ignore = true;
-        }
-        break;
-      case Node.COMMENT_NODE:
-        ignore = true;
-        break;
-      case Node.TEXT_NODE:
-        // If we capture CSSRules on style elements, ignore its text children nodes
-        // Since iterating over css rules can be 100X slower on certain browsers,
-        // we limit ignoring text nodes to STYLE elements with empty text content
-        let parent = node.parentNode as Element;
-        if (parent && parent.tagName === "STYLE" && parent.textContent.length === 0) {
-          ignore = true;
-        }
-      default:
-        break;
-    }
-  }
-  return ignore;
-}
-
-function shouldForceMaskNode(node: Node, parentForceMasked: boolean): boolean {
-  let mask = parentForceMasked;
-  if (!mask) {
-    if (node && node.nodeType === Node.ELEMENT_NODE) {
-      mask = (node as Element).getAttribute(ForceMaskAttribute) === "true";
-    }
-  }
-  return mask;
+function shouldUnmaskNode(node: Node, parentInfo: INodeInfo): boolean {
+  const parentUnmasked = parentInfo ? parentInfo.unmask : false;
```

Comment:
>unmask [](start = 49, length = 6)

do we guarantee that the parent's state is always set correctly before here or do we need to traverse the path to the root to make sure? Traversing the path and having up until the first parent that has an explicit value set seems "safer".

Reply:
The logic guarantees it, so unless there is a bug, parent will always be set and processed before child is processed. That said, with our default masking 'on', our risk in case of a bug now becomes just accidentally masking something extra, which wouldn't hurt us or users. Traversing all the way up the DOM for every node is definitely more straight-forward, but is also a perf nightmare.

---
In reply to: [260854704](https://github.com/Microsoft/clarity-js/pull/137#discussion_r260854704) [](ancestors = 260854704)